lib/stripe/stripe_client.rb

```diff
@@ -229,19 +229,36 @@ def format_app_info(info)
       str
     end
 
-    def handle_api_error(http_resp)
+    def handle_error_response(http_resp)
       begin
         resp = StripeResponse.from_faraday_hash(http_resp)
         error = resp.data[:error]
 
-        unless error && error.is_a?(Hash)
+        unless error
           raise StripeError.new("Indeterminate error")
         end
 
       rescue JSON::ParserError, StripeError
         raise general_api_error(http_resp[:status], http_resp[:body])
       end
 
+      if error.is_a?(Hash)
+        error = specific_api_error(resp, error)
+      elsif error.is_a?(String)
+        error = oauth_error(resp, error)
+      else
+        error = StripeError.new(
+          "Indeterminate error",
+          http_status: resp.http_status, http_body: resp.http_body,
+          json_body: resp.data, http_headers: resp.http_headers
+        )
+      end
```

Comment:
Ah, you're right.

Okay, what do you think about taking a slightly different approach here and recognizing specific OAuth errors and throwing typed versions of them? I think this is the direction that we should be trying to go anyway. It's a little riskier because a new OAuth error could fall into the wrong type until the library is updated, the OAuth API seems to be fairly stable at this point so I'm not sure that's a huge danger, and I like that by default errors will still be `APIError`. I wrote some code to support the idea below.

I'm also okay with making a server-side change to add some more distinctive identifier (if we're doing that anyway I'd suggest just adding a new field instead of using the `error_description` thing which was a little hacky to being with).

``` patch
diff --git a/lib/stripe/errors.rb b/lib/stripe/errors.rb
index c20d5e5..0ebe125 100644
--- a/lib/stripe/errors.rb
+++ b/lib/stripe/errors.rb
@@ -90,16 +90,43 @@ module Stripe
   class RateLimitError < StripeError
   end
 
-  # OAuthError is raised when the OAuth API returns an error.
-  class OAuthError < StripeError
-    attr_accessor :type
+  module OAuth
+    # OAuthError is raised when the OAuth API returns an error.
+    class OAuthError < StripeError
+      attr_accessor :code
 
-    def initialize(type, description, http_status: nil, http_body: nil, json_body: nil,
-                   http_headers: nil)
-      description ||= type
-      super(description, http_status: http_status, http_body: http_body,
-        json_body: json_body, http_headers: http_headers)
-      @type = type
+      def initialize(code, description, http_status: nil, http_body: nil, json_body: nil,
+                     http_headers: nil)
+        super(description, http_status: http_status, http_body: http_body,
+          json_body: json_body, http_headers: http_headers)
+        @code = code
+      end
+    end
+
+    # InvalidGrantError is raised when a specified code doesn't exist, is
+    # expired, has been used, or doesn't belong to you; a refresh token doesn't
+    # exist, or doesn't belong to you; or if an API key's mode (live or test)
+    # doesn't match the mode of a code or refresh token.
+    class InvalidGrantError < OAuthError
+    end
+
+    # InvalidRequestError is raised when a code, refresh token, or grant type
+    # parameter is not provided, but was required.
+    class InvalidRequestError < OAuthError
+    end
+
+    # InvalidScopeError is raised when an invalid scope parameter is provided.
+    class InvalidScopeError < OAuthError
+    end
+
+    # UnsupportedGrantTypeError is raised when an unuspported grant type
+    # parameter is specified.
+    class UnsupportedGrantTypeError < OAuthError
+    end
+
+    # UnsupportedResponseTypeError is raised when an unsupported response type
+    # parameter is specified.
+    class UnsupportedResponseTypeError < OAuthError
     end
   end
 end
diff --git a/lib/stripe/stripe_client.rb b/lib/stripe/stripe_client.rb
index 6b83ca3..2b20349 100644
--- a/lib/stripe/stripe_client.rb
+++ b/lib/stripe/stripe_client.rb
@@ -232,9 +232,9 @@ module Stripe
     def handle_error_response(http_resp)
       begin
         resp = StripeResponse.from_faraday_hash(http_resp)
-        error = resp.data[:error]
+        error_data = resp.data[:error]
 
-        unless error
+        unless error_data
           raise StripeError.new("Indeterminate error")
         end
 
@@ -242,57 +242,52 @@ module Stripe
         raise general_api_error(http_resp[:status], http_resp[:body])
       end
 
-      if error.is_a?(Hash)
-        error = specific_api_error(resp, error)
-      elsif error.is_a?(String)
-        error = oauth_error(resp, error)
-      else
-        error = StripeError.new(
-          "Indeterminate error",
-          http_status: resp.http_status, http_body: resp.http_body,
-          json_body: resp.data, http_headers: resp.http_headers
-        )
+      if error_data.is_a?(String)
+        error = specific_oauth_error(resp, error_data)
+      end
+      if error.nil?
+        error = specific_api_error(resp, error_data)
       end
 
       error.response = resp
       raise(error)
     end
 
-    def specific_api_error(resp, error)
+    def specific_api_error(resp, error_data)
       case resp.http_status
       when 400, 404
         error = InvalidRequestError.new(
-          error[:message], error[:param],
+          error_data[:message], error_data[:param],
           http_status: resp.http_status, http_body: resp.http_body,
           json_body: resp.data, http_headers: resp.http_headers
         )
       when 401
         error = AuthenticationError.new(
-          error[:message],
+          error_data[:message],
           http_status: resp.http_status, http_body: resp.http_body,
           json_body: resp.data, http_headers: resp.http_headers
         )
       when 402
         error = CardError.new(
-          error[:message], error[:param], error[:code],
+          error_data[:message], error_data[:param], error_data[:code],
           http_status: resp.http_status, http_body: resp.http_body,
           json_body: resp.data, http_headers: resp.http_headers
         )
       when 403
         error = PermissionError.new(
-          error[:message],
+          error_data[:message],
           http_status: resp.http_status, http_body: resp.http_body,
           json_body: resp.data, http_headers: resp.http_headers
         )
       when 429
         error = RateLimitError.new(
-          error[:message],
+          error_data[:message],
           http_status: resp.http_status, http_body: resp.http_body,
           json_body: resp.data, http_headers: resp.http_headers
         )
       else
         error = APIError.new(
-          error[:message],
+          error_data[:message],
           http_status: resp.http_status, http_body: resp.http_body,
           json_body: resp.data, http_headers: resp.http_headers
         )
@@ -301,21 +296,24 @@ module Stripe
       error
     end
 
-    def oauth_error(resp, error)
-      description = resp.data[:error_description] || error
-
-      if error == 'invalid_request'
-        error = InvalidRequestError.new(
-          description, nil,
-          http_status: resp.http_status, http_body: resp.http_body,
-          json_body: resp.data, http_headers: resp.http_headers
-        )
+    # Attempts to look at a response's error code and return an OAuth error if
+    # one matches. Will return `nil` if the code isn't recognized.
+    def specific_oauth_error(resp, error_code)
+      description = resp.data[:error_description] || error_code
+
+      args = [error_code, description, {
+        http_status: resp.http_status, http_body: resp.http_body,
+        json_body: resp.data, http_headers: resp.http_headers
+      }]
+
+      case error_code
+      when 'invalid_grant'             then OAuth::InvalidGrantError.new(*args)
+      when 'invalid_request'           then OAuth::InvalidRequestError.new(*args)
+      when 'invalid_scope'             then OAuth::InvalidScopeError.new(*args)
+      when 'unsupported_grant_type'    then OAuth::UnsupportedGrantTypeError.new(*args)
+      when 'unsupported_response_type' then OAuth::UnsupportedResponseTypeError.new(*args)
       else
-        error = OAuthError.new(
-          error, description,
-          http_status: resp.http_status, http_body: resp.http_body,
-          json_body: resp.data, http_headers: resp.http_headers
-        )
+        nil
       end
     end
 
diff --git a/test/stripe/stripe_client_test.rb b/test/stripe/stripe_client_test.rb
index 9b02c12..4fdc696 100644
--- a/test/stripe/stripe_client_test.rb
+++ b/test/stripe/stripe_client_test.rb
@@ -342,7 +342,7 @@ module Stripe
 
           client = StripeClient.new
           opts = {api_base: Stripe.connect_base}
-          e = assert_raises Stripe::InvalidRequestError do
+          e = assert_raises Stripe::OAuth::InvalidRequestError do
             client.execute_request(:post, '/oauth/token', opts)
           end
 
@@ -360,12 +360,12 @@ module Stripe
 
           client = StripeClient.new
           opts = {api_base: Stripe.connect_base}
-          e = assert_raises Stripe::OAuthError do
+          e = assert_raises Stripe::OAuth::InvalidGrantError do
             client.execute_request(:post, '/oauth/token', opts)
           end
 
           assert_equal(400, e.http_status)
-          assert_equal('invalid_grant', e.type)
+          assert_equal('invalid_grant', e.code)
           assert_equal('This authorization code has already been used. All tokens issued with this code have been revoked.', e.message)
         end
       end
```